### PR TITLE
feat: Integrate cohere models

### PR DESCRIPTION
**Pull request by Wendong-Fan**
## Description

Integrate cohere models

contributed by @NeilJohnson0930 

## Motivation and Context

Why is this change required? What problem does it solve?
If it fixes an open issue, please link to the issue here.
You can use the syntax `close https://github.com/camel-ai/camel/issues/896'

- [ ] I have raised an issue to propose this change ([required](https://github.com/camel-ai/camel/issues) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Put an `x` in all the boxes that apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [ ] Example (update in the folder of example)

## Implemented Tasks

- [x] Develop the script to interact with the Cohere model API.
- [x] Ensure Cohere model is seamlessly integrated within the Camel framework.
- [x] Write and execute unit tests to verify the functionality of the Cohere model integration.
- [x] Resolve compatibility issues between the Cohere application and the retriever component.

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [ ] I have read the [CONTRIBUTION](https://github.com/camel-ai/camel/blob/master/CONTRIBUTING.md) guide. (**required**)
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly. (*required for a bug fix or a new feature*)
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: camel/models/cohere_model.py [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+
+        self._client = cohere.Client(api_key=self._api_key)
+        self._token_counter: Optional[BaseTokenCounter] = None
+
+    def _to_openai_response(
+        self, response: 'NonStreamedChatResponse'
+    ) -> ChatCompletion:
+        unique_id = str(uuid.uuid4())
+
+        # Safely access nested attributes
```
Comment (koch3092) [inline camel/models/cohere_model.py:69]: Regarding the operation of structured data, how about using BaseModel?

File: examples/models/role_playing_with_cohere.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+from colorama import Fore
+
+from camel.configs import CohereConfig
+from camel.models import ModelFactory
+from camel.societies import RolePlaying
+from camel.types import ModelPlatformType, ModelType
+from camel.utils import print_text_animated
+
+
+def test_cohere_api(api_key):
```
Comment (koch3092) [inline examples/models/role_playing_with_cohere.py:27]: Just curious, I saw that assert was not used in the test. Is it because the results are unpredictable?

**Review by koch3092 — COMMENTED** [review]
Thanks @NeilJohnson0930 , left some comments

File: camel/configs/cohere_config.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+            values closer to 1 make it more random. Defaults to None.
+        p (Optional[float]): Sets a p% nucleus sampling threshold.
+            Defaults to None.
+        k (Optional[int]): Limits the number of tokens to sample from on
+            each step. Defaults to None.
+        max_tokens (Optional[int]): The maximum number of tokens to generate.
+            Defaults to None.
+        prompt_truncation (Optional[str]): How to truncate the prompt if it
+            exceeds the model's context length. Can be 'START', 'END', or
+            'AUTO'. Defaults to None.
```
Comment (WHALEEYE) [inline camel/configs/cohere_config.py:39]:
```suggestion
        temperature (Optional[float], optional): Controls randomness in the model.
            Values closer to 0 make the model more deterministic, while
            values closer to 1 make it more random. (default: :obj:`None`)
        p (Optional[float], optional): Sets a p% nucleus sampling threshold.
            (default: :obj:`None`)
        k (Optional[int], optional): Limits the number of tokens to sample from on
            each step. (default: :obj:`None`)
        max_tokens (Optional[int], optional): The maximum number of tokens to generate.
            (default: :obj:`None`)
        prompt_truncation (Optional[str], optional): How to truncate the prompt if it
            exceeds the model's context length. Can be 'START', 'END', or
            'AUTO'. (default: :obj:`None`)
```

File: camel/configs/cohere_config.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    prompt_truncation: Optional[str] = None
+
+    @field_validator("prompt_truncation")
+    @classmethod
+    def validate_prompt_truncation(cls, v):
+        if v is not None and v not in ['START', 'END', 'AUTO']:
+            raise ValueError(
+                "prompt_truncation must be 'START', 'END', or 'AUTO'"
+            )
+        return v
```
Comment (WHALEEYE) [inline camel/configs/cohere_config.py:55]: I think we can just use `Optional[Literal['START', 'END', 'AUTO']]` for `prompt_truncation` instead of doing this checking?

File: camel/models/cohere_model.py [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+    if os.getenv("AGENTOPS_API_KEY") is not None:
+        from agentops import LLMEvent, record
+    else:
+        raise ImportError
+except (ImportError, AttributeError):
+    LLMEvent = None
+
+
+class CohereModel(BaseModelBackend):
+    """Cohere API in a unified BaseModelBackend interface."""
```
Comment (Wendong-Fan) [inline camel/models/cohere_model.py:43]:
```suggestion
    r"""Cohere API in a unified BaseModelBackend interface."""
```

File: camel/models/cohere_model.py [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+    def __init__(
+        self,
+        model_type: ModelType,
+        model_config_dict: Dict[str, Any],
+        api_key: Optional[str] = None,
+        token_counter: Optional[BaseTokenCounter] = None,
+        model_platform: Optional[ModelPlatformType] = None,
+    ):
+        import cohere
+
```
Comment (Wendong-Fan) [inline camel/models/cohere_model.py:54]: missing docstring

File: camel/models/cohere_model.py [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+class CohereModel(BaseModelBackend):
+    """Cohere API in a unified BaseModelBackend interface."""
+
+    def __init__(
+        self,
+        model_type: ModelType,
+        model_config_dict: Dict[str, Any],
+        api_key: Optional[str] = None,
+        token_counter: Optional[BaseTokenCounter] = None,
+        model_platform: Optional[ModelPlatformType] = None,
```
Comment (Wendong-Fan) [inline camel/models/cohere_model.py:51]: what's this `model_platform` used for? Missing `url`

File: camel/models/cohere_model.py [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+                new_messages.append({"role": "System", "message": content})
+            else:
+                raise ValueError(f"Unsupported message role: {role}")
+
+        return new_messages  # type: ignore[return-value]
+
+    @property
+    def token_counter(self) -> BaseTokenCounter:
+        """Initialize the token counter for the model backend.
+
```
Comment (Wendong-Fan) [inline camel/models/cohere_model.py:147]:
```suggestion
        r"""Initialize the token counter for the model backend.

```

File: camel/models/cohere_model.py [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+        Returns:
+            BaseTokenCounter: The token counter following the model's
+                tokenization style.
+        """
+        if not self._token_counter:
+            self._token_counter = OpenAITokenCounter(
+                model=ModelType.GPT_3_5_TURBO
```
Comment (Wendong-Fan) [inline camel/models/cohere_model.py:154]: use GPT_4O_MINI as default as GPT_3_5_TURBO will be depressed

```suggestion
                model=ModelType.GPT_4O_MINI
```

File: camel/models/cohere_model.py [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+            )
+        return self._token_counter
+
+    @api_keys_required("COHERE_API_KEY")
+    def run(self, messages: List[OpenAIMessage]) -> ChatCompletion:
+        """Runs inference of Cohere chat completion.
```
Comment (Wendong-Fan) [inline camel/models/cohere_model.py:160]:
```suggestion
        r"""Runs inference of Cohere chat completion.
```

File: camel/models/cohere_model.py [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+        cohere_messages = self._to_cohere_chatmessage(messages)
+
+        # Filter out unsupported parameters
+        supported_params = {
+            'temperature',
+            'p',
+            'k',
+            'max_tokens',
+            'prompt_truncation',
+        }
```
Comment (Wendong-Fan) [inline camel/models/cohere_model.py:180]: this part of check is done by `self.check_model_config()` in `BaseModelBackend`

File: camel/models/cohere_model.py [hunk 12]
```diff
@@ -0,0 +1,240 @@
...
+        for param in self.model_config_dict:
+            if param not in supported_params:
+                raise ValueError(
+                    f"Unexpected argument `{param}` is "
+                    "input into Cohere model backend."
+                )
```
Comment (Wendong-Fan) [inline camel/models/cohere_model.py:240]: use COHERE_API_PARAMS in `camel/configs/cohere_config.py` to do the check

File: examples/models/cohere_model_example.py [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+
+user_msg = BaseMessage.make_user_message(
+    role_name="User",
+    content="""Say hi to CAMEL AI, one open-source community dedicated to the 
+    study of autonomous and communicative agents.""",
+)
+
+# Get response information
+response = camel_agent.step(user_msg)
+print(response.msgs[0].content)
```
Comment (Wendong-Fan) [inline examples/models/cohere_model_example.py:43]: add the printed output in below

File: examples/models/role_playing_with_cohere.py (hunk 2 above)
Comment (Wendong-Fan) [reply examples/models/role_playing_with_cohere.py:27]: thanks @koch3092 , this is not test but an example, but we indeed missing test file, @NeilJohnson0930 you can refer to test code in test/models/test_mistral_model.py

**Review by Wendong-Fan — COMMENTED** [review]
Thanks @NeilJohnson0930 ! Left some comments below

File: camel/models/cohere_model.py (hunk 1 above)
Comment (NeilJohnson0930) [reply camel/models/cohere_model.py:69]: PR was developed in Basemodel framework, ChatCompletion.construct needs a parameter "id" which cannot receive from cohere api response. Uuid.uuid4() was simulated  'id' as a parameter passed to basemodel function, which is used in other parts of the project.

File: examples/models/role_playing_with_cohere.py (hunk 2 above)
Comment (NeilJohnson0930) [reply examples/models/role_playing_with_cohere.py:27]: Got it! test_cohere_model.py will be committed as soon as possible

File: camel/configs/cohere_config.py (hunk 3 above)
Comment (NeilJohnson0930) [reply camel/configs/cohere_config.py:39]: Thanks WHALEEYE！Your comment will be updated on new commit.

File: camel/configs/cohere_config.py (hunk 4 above)
Comment (NeilJohnson0930) [reply camel/configs/cohere_config.py:55]: You're right, it's more appropriate for the base model to catch errors related to overly long prompts by default. check will be offlined in new commit.

File: camel/models/cohere_model.py (hunk 6 above)
Comment (NeilJohnson0930) [reply camel/models/cohere_model.py:54]: Thanks bro! docstring will be added in new commit.

File: camel/models/cohere_model.py (hunk 7 above)
Comment (NeilJohnson0930) [reply camel/models/cohere_model.py:51]: Hi Wendong,

The model_platform defined here is used to select a specific Cohere model. For example:
`
elif model_platform.is_cohere and model_type.is_cohere:
    model_class = CohereModel
`
Additionally, the URL is not required when making API calls to Cohere, as indicated in the official documentation. For example:
`
import cohere
co = cohere.Client(api_key="<YOUR API KEY>")
response = co.chat(
  model="command-r-plus-08-2024",
  message="Write a title for a blog post about API design. Only output the title text."
)
`

File: camel/models/cohere_model.py (hunk 11 above)
Comment (NeilJohnson0930) [reply camel/models/cohere_model.py:180]: Got it. detele in new commit

File: camel/models/cohere_model.py (hunk 12 above)
Comment (NeilJohnson0930) [reply camel/models/cohere_model.py:240]: Hi Wendong, I need to confirm if the function check_model_config should be implemented in camel/configs/cohere_config.py to perform the check, 
as I noticed that the checks for other models are implemented in camel/models/*_model.py.

File: examples/models/cohere_model_example.py (hunk 13 above)
Comment (NeilJohnson0930) [reply examples/models/cohere_model_example.py:43]: I'm confused here. Print is already on the bottom line of file cohere_model_example.py.

File: camel/configs/cohere_config.py [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+        k (Optional[int], optional):
+            Limits the number of tokens to sample from on
+            each step. (default: :obj:`None`)
+        max_tokens (Optional[int], optional):
+            The maximum number of tokens to generate.
+            (default: :obj:`None`)
+        prompt_truncation (Optional[str], optional):
+            How to truncate the prompt if it
+            exceeds the model's context length. Can be 'START', 'END', or
+            'AUTO'. (default: :obj:`None`)
```
Comment (Wendong-Fan) [inline camel/configs/cohere_config.py:40]: from https://docs.cohere.com/reference/chat seems not all parameters are included in config, like `seed `

File: camel/models/cohere_model.py [hunk 15]
```diff
@@ -0,0 +1,216 @@
...
+                        safe_get(
+                            response,
+                            'meta',
+                            'billed_tokens',
+                            'completion_tokens',
+                        )
+                        or 0
+                    )
+                ),
+            },
```
Comment (Wendong-Fan) [inline camel/models/cohere_model.py:122]: the path we get usage data seems not correct, it should from `response.meta.tokens.input_tokens` and `response.meta.tokens.output_tokens`

File: camel/models/cohere_model.py [hunk 16]
```diff
@@ -0,0 +1,216 @@
...
+        )
+        self._api_key = api_key or os.environ.get("COHERE_API_KEY")
+        self._url = url or os.environ.get("COHERE_SERVER_URL")
+
+        self._client = cohere.Client(api_key=self._api_key)
+        self._token_counter: Optional[BaseTokenCounter] = None
+
+    def _to_openai_response(
+        self, response: 'NonStreamedChatResponse'
+    ) -> ChatCompletion:
```
Comment (Wendong-Fan) [inline camel/models/cohere_model.py:67]: cohere model also support tool calling, we need to add tool_calls data to `ChatCompletion `

File: camel/models/cohere_model.py [hunk 17]
```diff
@@ -0,0 +1,216 @@
...
+            content = msg.get("content")
+
+            if role == "user":
+                new_messages.append({"role": "User", "message": content})
+            elif role == "assistant":
+                new_messages.append({"role": "Chatbot", "message": content})
+            elif role == "system":
+                new_messages.append({"role": "System", "message": content})
+            else:
+                raise ValueError(f"Unsupported message role: {role}")
```
Comment (Wendong-Fan) [inline camel/models/cohere_model.py:141]: since cohere's `message ` is just pure string, we can just pass content and remove the role information

**Review by Wendong-Fan — COMMENTED** [review]
Thanks @NeilJohnson0930 ! Left some comments

File: camel/models/cohere_model.py [hunk 18]
```diff
@@ -0,0 +1,276 @@
...
+            + (response.usage.tokens.output_tokens or 0),  # type: ignore[union-attr]
+        }
+
+        tool_calls = response.message.tool_calls  # type: ignore[union-attr]
+        choices = []
+        if tool_calls:
+            for tool_call in tool_calls:
+                openai_tool_calls = [
+                    dict(
+                        id=tool_call.id,  # type: ignore[union-attr]
```
Comment (Wendong-Fan) [inline camel/models/cohere_model.py]: type ignore could be removed

File: camel/models/cohere_model.py [hunk 19]
```diff
@@ -0,0 +1,276 @@
...
+                        function={
+                            "name": tool_call.function.name,  # type: ignore[union-attr]
+                            "arguments": tool_call.function.arguments,  # type: ignore[union-attr]
```
Comment (Wendong-Fan) [inline camel/models/cohere_model.py]: use if statement instead of type ignore

File: camel/models/cohere_model.py [hunk 20]
```diff
@@ -0,0 +1,276 @@
...
+                        },
+                        type=tool_call.type,
+                    )
+                ]
+
+                choice = dict(
+                    index=None,
+                    message={
+                        "role": "assistant",
+                        "content": response.message.tool_plan,  # type: ignore[union-attr]
```
Comment (Wendong-Fan) [inline camel/models/cohere_model.py]: type ignore could be removed

File: camel/models/cohere_model.py [hunk 21]
```diff
@@ -0,0 +1,276 @@
...
+
+        new_messages = []
+        for msg in messages:
+            role = msg.get("role")
+            content = msg.get("content")
+            function_call = msg.get("function_call")
+
+            if role == "user":
+                new_message = UserChatMessageV2(role="user", content=content)  # type: ignore[arg-type]
+            elif role == "function":
```
Comment (Wendong-Fan) [inline camel/models/cohere_model.py]: should also support `tool`

File: camel/models/cohere_model.py [hunk 22]
```diff
@@ -0,0 +1,276 @@
...
+                            id=assis_tool_call_id,
+                            type="function",
+                            function=ToolCallV2Function(
+                                name=function_call.get("name"),  # type: ignore[attr-defined]
+                                arguments=arguments_json,  # type: ignore[attr-defined]
+                            ),
+                        )
+                    ]
+                    if function_call
+                    else None,
```
Comment (Wendong-Fan) [inline camel/models/cohere_model.py]: this is duplicated check, can be removed

File: camel/toolkits/math_toolkit.py [hunk 23]
```diff
@@ -47,7 +47,7 @@ def sub(self, a: int, b: int) -> int:
         Returns:
             integer: The result of subtracting :obj:`b` from :obj:`a`.
         """
-        return a - b
+        return int(a) - int(b)
```
Comment (Wendong-Fan) [inline camel/toolkits/math_toolkit.py]: unnecessary change

File: camel/types/enums.py [hunk 24]
```diff
@@ -89,6 +89,12 @@ class ModelType(UnifiedModelType, Enum):
     REKA_FLASH = "reka-flash"
     REKA_EDGE = "reka-edge"
 
+    # Cohere models
+    COHERE_COMMAND_R = "command-r"
+    COHERE_COMMAND_LIGHT = "command-light"
+    COHERE_COMMAND = "command"
+    COHERE_COMMAND_NIGHTLY = "command-nightly"
```
Comment (Wendong-Fan) [inline camel/types/enums.py:97]: from https://docs.cohere.com/v2/docs/models more models like `command-r-plus` could be supported

File: camel/types/enums.py [hunk 25]
```diff
@@ -389,6 +409,13 @@ def token_limit(self) -> int:
             ModelType.GEMINI_EXP_1114,  # Not given in docs, assuming the same
         }:
             return 1_048_576
+        elif self in {
+            ModelType.COHERE_COMMAND_R,
+            ModelType.COHERE_COMMAND_LIGHT,
+            ModelType.COHERE_COMMAND,
+            ModelType.COHERE_COMMAND_NIGHTLY,
+        }:
+            return 4_096
```
Comment (Wendong-Fan) [inline camel/types/enums.py]: we have one dict for 4_096, for COHERE_COMMAND_R it should be 128k

File: examples/models/cohere_model_example.py [hunk 26]
```diff
@@ -0,0 +1,51 @@
...
+from camel.types import ModelPlatformType, ModelType
+
+model = ModelFactory.create(
+    model_platform=ModelPlatformType.COHERE,
+    model_type=ModelType.COHERE_COMMAND_R,
+    model_config_dict=CohereConfig(temperature=0.0).as_dict(),
+)
+
+# Define system message
+sys_msg = BaseMessage.make_assistant_message(
```
Comment (Wendong-Fan) [inline examples/models/cohere_model_example.py]: message setting could be simpified

File: camel/models/cohere_model.py [hunk 27]
```diff
@@ -0,0 +1,276 @@
...
+                    tool_call_id=self.tool_call_id,  # type: ignore[arg-type]
+                    content=content,  # type: ignore[assignment,arg-type]
+                )
+            elif role == "assistant":
+                if function_call is None:
+                    new_message = AssistantChatMessageV2(  # type: ignore[assignment]
+                        role="assistant",
+                        content=content,  # type: ignore[arg-type]
+                    )
+                    continue
```
Comment (Wendong-Fan) [inline camel/models/cohere_model.py]: I think here we should use `else` statement instead of break the whole loop

File: camel/models/cohere_model.py [hunk 28]
```diff
@@ -0,0 +1,276 @@
...
+                arguments = function_call.get("arguments")  # type: ignore[attr-defined]
+                arguments_dict = ast.literal_eval(arguments)
+                arguments_json = json.dumps(arguments_dict)
+                assis_tool_call_id = str(uuid.uuid4())
+                self.tool_call_id = assis_tool_call_id  # type: ignore[assignment]
```
Comment (Wendong-Fan) [inline camel/models/cohere_model.py]: unnecessary type ignore

**Review by Wendong-Fan — COMMENTED** [review]
Thanks @NeilJohnson0930 , left some comments below, based on the review also added one more commit here: feel free to check~
